### PR TITLE
RSDK-4171, RSDK-4546 - Fix weird module behavior on process kill and model switch

### PR DESCRIPTION
**Pull request by cheukt**
If we kill a module process, the viam-server starts the module back up but the WeakDependents are not updated - so in cases where the module now serves a different model, certain things will not get updated (sensors, motion, etc). Causes issues when you have a sensor that no longer exists showing up on the sensors service card. (not sure of a good test for this, open to suggestions!)

if we switch the model of a modular resource to an invalid model and then back, we were also unable to reinstantiate the resource - we closed the old one but because the invalid model never succeeded, on reconfiguring it back to the valid model the rdk recognizes it as a regular reconfigure. Since the old resource is already gone, the module will respond with various kinds of ResourceNotFound errors - this change correctly unsets the resource from the graph node.

### CODE REVIEW

**Review by benjirewis — COMMENTED** [review]
Thanks sm for looking into this! I'll think about ways to test that weak dependents issue; I _think_ it's possible.

**Review by stuqdog — APPROVED** [review]

File: resource/graph_node.go [hunk 1]
```diff
@@ -119,6 +119,14 @@ func (w *GraphNode) IsUninitialized() bool {
 	return w.current == nil
 }
 
+// UnsetResource unsets the current resource. This function does not clean up
+// or close the resource and should be used carefully.
+func (w *GraphNode) UnsetResource() {
+	w.mu.RLock()
```
[the remark this answers is not in the document]
Comment (cheukt) [reply resource/graph_node.go]: fixed, thanks for catching!

File: module/modmanager/manager.go [hunk 2]
```diff
@@ -493,6 +493,7 @@ func (mgr *Manager) newOnUnexpectedExitHandler(mod *module) func(exitCode int) b
 			if _, err := mgr.addResource(ctx, res.conf, res.deps); err != nil {
 				mgr.logger.Warnw("error while re-adding resource to module",
 					"resource", name, "module", mod.name, "error", err)
+				delete(mgr.rMap, name)
```
Comment (benjirewis) [inline module/modmanager/manager.go:496]: Actually I think this resolves RSDK-4546 @cheukt.

**Review by benjirewis — APPROVED** [review]
Apologies for delay; was trying to write a test that a resource that is weakly dependent on a modular resource that gets orphaned will get reconfigured. It was.. difficult, so I will certainly not block the PR on writing one unless you're feeling inspired @cheukt. This LGTM and thanks for the changes. Could we also add RSDK-4546 to the PR title?

**Pull request merged**